phoenix-core/src/it/java/org/apache/phoenix/end2end/WALAnnotationIT.java

```diff
@@ -0,0 +1,550 @@
+            table = PhoenixRuntime.getTableNoCache(conn, builder.getEntityTableName());
+            Assert.assertTrue("Change detection is disabled when it should be enabled!",
+                table.isChangeDetectionEnabled());
+            //set to FALSE
+            String disableSql =
+                "ALTER TABLE " + builder.getEntityTableName() +
+                    " SET " + CHANGE_DETECTION_ENABLED + "=FALSE";
+            conn.createStatement().execute(disableSql);
+            table = PhoenixRuntime.getTableNoCache(conn, builder.getEntityTableName());
+            Assert.assertFalse("Change detection is enabled when it should be disabled!",
+                table.isChangeDetectionEnabled());
+            //now upsert again
+            conn.createStatement().execute(upsertSql);
+            //check that we still didn't annotate anything
+            entries = getEntriesForTable(TableName.valueOf(builder.getPhysicalTableName(false)));
+            Assert.assertEquals(0, entries.size());
+        }
+    }
+
+    @Test
+    public void testCantSetChangeDetectionOnIndex() throws Exception {
+        Assume.assumeTrue(HbaseCompatCapabilities.hasPreWALAppend());
+        try (Connection conn = DriverManager.getConnection(getUrl())) {
+            SchemaBuilder builder = new SchemaBuilder(getUrl());
+            builder.withTableDefaults().build();
+            try {
+                String badIndexSql =
+                    "CREATE INDEX IDX_SHOULD_FAIL"  + " ON " + builder.getEntityTableName() +
+                        "(COL1) "
+                        + CHANGE_DETECTION_ENABLED + "=TRUE";
+                conn.createStatement().execute(badIndexSql);
+                Assert.fail("Didn't throw a SQLException for setting change detection on an " +
+                    "index at create time!");
+            } catch (SQLException se) {
+                TestUtil.assertSqlExceptionCode(
+                    SQLExceptionCode.CHANGE_DETECTION_SUPPORTED_FOR_TABLES_AND_VIEWS_ONLY, se);
+            }
+        }
+    }
+
+    @Test
+    public void testUpsertAndDeleteWithGlobalIndex() throws Exception {
+        Assume.assumeTrue(HbaseCompatCapabilities.hasPreWALAppend());
+        SchemaBuilder builder = new SchemaBuilder(getUrl());
+        boolean createGlobalIndex = true;
+        long ddlTimestamp = upsertAndDeleteHelper(builder, createGlobalIndex);
+        assertAnnotation(2, builder.getPhysicalTableName(false), null,
+            builder.getTableOptions().getSchemaName(),
+            builder.getDataOptions().getTableName(), PTableType.TABLE, ddlTimestamp);
+        assertAnnotation(0, builder.getPhysicalTableIndexName(false),
+            null, builder.getTableOptions().getSchemaName(),
+            SchemaUtil.getTableNameFromFullName(builder.getEntityTableIndexName()),
+            PTableType.INDEX,
+            ddlTimestamp);
+    }
+
+    //Note that local secondary indexes aren't supported because they go in the same WALEdit as the
```

Comment:
Actually the opposite -- when a local index is present I believe the WALEdits will have a mix of Cells from both the base table and the local index, because local indexes are stored within the actual table, just in a separate CF. So the later phases of change detection (that read the WALs) will have to filter out Cells that have the local index CF.

Reply:
Ok, got it.